jar-infer/jar-infer-lib/src/main/java/com/uber/nullaway/jarinfer/BytecodeAnnotator.java

```diff
@@ -46,77 +48,52 @@ private static void LOG(boolean cond, String tag, String msg) {
   public static final String javaxNullableDesc = "Ljavax/annotation/Nullable;";
   public static final String javaxNonnullDesc = "Ljavax/annotation/Nonnull;";
 
-  private final MethodParamAnnotations nullableParams;
-  private final MethodReturnAnnotations nullableReturns;
-
-  private String className = "";
-
-  public BytecodeAnnotator(
-      ClassVisitor cv,
-      MethodParamAnnotations nullableParams,
-      MethodReturnAnnotations nullableReturns) {
-    super(ASM7, cv);
-    this.nullableParams = nullableParams;
-    this.nullableReturns = nullableReturns;
-  }
-
-  @Override
-  public void visit(
-      final int version,
-      final int access,
-      final String name,
-      final String signature,
-      final String superName,
-      final String[] interfaces) {
-    className = name.replace('/', '.');
-    if (cv != null) {
-      cv.visit(version, access, name, signature, superName, interfaces);
+  private static void addAnnotationIfNotPresent(
+      List<AnnotationNode> annotationList, String annotation) {
+    for (AnnotationNode node : annotationList) {
+      if (node.desc.equals(annotation)) {
+        return;
+      }
     }
+    annotationList.add(new AnnotationNode(Opcodes.ASM7, annotation));
   }
 
-  @Override
-  public MethodVisitor visitMethod(
-      final int access,
-      final String name,
-      final String desc,
-      final String signature,
-      final String[] exceptions) {
-    MethodVisitor mv = cv.visitMethod(access, name, desc, signature, exceptions);
-    LOG(debug, "DEBUG", "ClassName: " + className);
-    LOG(debug, "DEBUG", "visited method - " + name + " desc: " + desc + " sign: " + signature);
-    String methodSignature = className + "." + name + desc;
-    if (mv != null) {
+  private static void annotateBytecode(
+      InputStream is,
+      OutputStream os,
+      MethodParamAnnotations nonnullParams,
+      MethodReturnAnnotations nullableReturns)
+      throws IOException {
+    ClassReader cr = new ClassReader(is);
+    ClassWriter cw = new ClassWriter(0);
+    ClassNode cn = new ClassNode(Opcodes.ASM7);
+    cr.accept(cn, 0);
+
+    String className = cn.name.replace('/', '.');
+    List<MethodNode> methods = cn.methods;
+    for (MethodNode method : methods) {
+      String methodSignature = className + "." + method.name + method.desc;
       if (nullableReturns.contains(methodSignature)) {
         // Add a @Nullable annotation on this method to indicate that the method can return null.
-        mv.visitAnnotation(javaxNullableDesc, true);
+        addAnnotationIfNotPresent(method.visibleAnnotations, javaxNullableDesc);
```

Comment:
So, this generally makes sense to me, but I am not sure why we need to switch the API. I thought one strength of the design was that we were using one API to do the modification and another in our tests?

Wouldn't it have been an easier change to subclass `MethodVisitor` and keep just enough state to drop repeated values from calls to `visitAnnotation` and `visitParameterAnnotation`? Although, maybe this is still cleaner than that? I can go either way on this, but I remember you pushed for the separate API for main and test code model. 😄

Reply:
I felt that using the visitor API for this by saving the state to filter repeated values is unnecessarily complicated. In comparison, the tree based API is very simple and easy to understand.

Also, that argument of different APIs was more when we were just looking at whether the updated class has the list of annotations we were looking for (so, the test code would have been essentially the same as this code in that case). Now that we have tests using the 'Expect*' annotations as well as the entries comparison in Jars / Aars, I feel confident about have the same API in this code as well as the test. 

If you still feel strongly about having different APIs, I will try to use the visitor API here.